src/Microsoft.DncEng.SecretManager/ParameterConverter.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.DncEng.SecretManager
+{
+    public static class ParameterConverter
+    {
+        public static TParameters ConvertParameters<TParameters>(IDictionary<string, object> parameters)
+        {
+            return (TParameters)ConvertValue(parameters, typeof(TParameters));
+        }
+
+        private static object ConvertValue(object value, Type type)
+        {
+            if (IsDictionary(value))
+            {
+                return ConvertObject(value, type);
+            }
+
+            return ConvertScalar(value, type);
+        }
+
+        private static bool IsDictionary(object value)
+        {
+            var type = value.GetType();
+            if (type.GetInterfaces().Any(iface => iface.Name.Contains("Dictionary")))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static object ConvertObject(object value, Type type)
+        {
+            var ciDict = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+            foreach (dynamic entry in (IEnumerable) value)
```

Comment:
dynamic...? Can we not just make it IReadOnlyDictionary? Which classes don't inherit from that that we want to handle?

Reply:
This isn't for IReadOnlyDictionary, some of the types are IDictionary<object, object> and others are IDictionary<string, object>. Its way more code that is less easy to read with more reflection to handle multiple types for keys.

Reply:
Ah, yeah, there isn't an IReadOnlyDictionary (non generic).

Reply:
nope, and one of the types doesn't even inplement non-generic IDictionary, so I couldn't use that.

Reply:
I feel the dynamic really expresses the intention here, "its a list of things that have keys and values", its probably a dictionary, but if its a HashTable or something else that fine too.

Reply:
I think the equivalent without dynamic would be like 4 methods with lots of reflection, which I could write, but I don't feel it adds much value.

Reply:
if only IReadOnlyDictionary and co/contravariance were properly supported. Then this method could accept IReadOnlyDictionary<object, object> and an IReadOnlyDictionary<string, object> could be passed to it.

Reply:
I agree. It's an internal took, and I hadn't thought about the type variance. The dynamic isn't worth it. :-)